include/alp/omp/blas3.hpp

```diff
@@ -0,0 +1,526 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * @author A. N. Yzelman
+ * @date 14th of January 2022
+ */
+
+#ifndef _H_ALP_OMP_BLAS3
+#define _H_ALP_OMP_BLAS3
+
+#include <algorithm>   // for std::min/max
+#include <type_traits> // for std::enable_if
+
+#include <alp/descriptors.hpp>
+#include <alp/structures.hpp>
+
+// #include <alp/amf-based/matrix.hpp>
+#include <alp/base/blas3.hpp>
+
+#include "matrix.hpp"
+#include "storage.hpp"
+
+// Include backend to which sequential work is delegated
+#ifdef _ALP_OMP_WITH_REFERENCE
+ #include <alp/reference/blas2.hpp>
+ #include <alp/reference/blas3.hpp>
+ #include <alp/reference/io.hpp>
+#endif
+
+#ifndef NDEBUG
+#include "../../../tests/utils/print_alp_containers.hpp"
+#endif
+
+
+namespace alp {
+
+	namespace internal {
+
+		/**
+		 * \internal general mxm implementation that all mxm variants using 
+		 * structured matrices refer to.
+		 */
+		template<
+			bool allow_void,
+			class MulMonoid,
+			typename OutputType, typename InputType1, typename InputType2,
+			class Operator, class Monoid,
+			typename OutputStructure, typename OutputView, 
+			typename OutputImfR, typename OutputImfC,
+			typename InputStructure1, typename InputView1, 
+			typename InputImfR1, typename InputImfC1,
+			typename InputStructure2, typename InputView2, 
+			typename InputImfR2, typename InputImfC2
+		>
+		RC mxm_generic( 
+			alp::Matrix< OutputType, OutputStructure, 
+			Density::Dense, OutputView, OutputImfR, OutputImfC, omp > &C,
+			alp::Matrix< InputType1, InputStructure1, 
+			Density::Dense, InputView1, InputImfR1, InputImfC1, omp > &A,
+			alp::Matrix< InputType2, InputStructure2, 
+			Density::Dense, InputView2, InputImfR2, InputImfC2, omp > &B,
+			const Operator &oper,
+			const Monoid &monoid,
+			const MulMonoid &mulMonoid,
+			const typename std::enable_if< !alp::is_object< OutputType >::value &&
+				!alp::is_object< InputType1 >::value && !
+				alp::is_object< InputType2 >::value &&
+				alp::is_operator< Operator >::value &&
+				alp::is_monoid< Monoid >::value,
+			void >::type * const = NULL
+		) {
+
+			static_assert( 
+				!(
+					std::is_same< InputType1, void >::value ||
+					std::is_same< InputType2, void >::value
+				),
+				"alp::internal::mxm_generic: the operator-monoid version of mxm cannot be "
+				"used if either of the input matrices is a pattern matrix (of type "
+				"void)"
+			);
+
+#ifndef NDEBUG
+			std::cout << "In alp::internal::mxm_generic (omp)\n";
+#endif
+
+			// Early exit checks 
+			if( ! internal::getInitialized( A ) ||
+				! internal::getInitialized( B ) ||
+				! internal::getInitialized( C )
+			) {
+				internal::setInitialized( C, false );
+				return SUCCESS;
+			}
+
+			const std::ptrdiff_t m   { static_cast< std::ptrdiff_t >( nrows( C ) ) };
```

Comment:
I think that the current style recommends only one white space between `m   {` , 
the same comment for the lines below.